torch/quantization/_equalize.py

```diff
@@ -0,0 +1,130 @@
+    bias = module1.bias
+
+    weight1_range = channel_range(weight1, output_axis)
+    weight2_range = channel_range(weight2, input_axis)
+
+    # producing scaling factors to applied
+    weight2_range += 1e-9
+    scaling_factors = torch.sqrt(weight1_range / weight2_range)
+    inverse_scaling_factors = torch.reciprocal(scaling_factors)
+
+    bias = bias * inverse_scaling_factors
+
+    # formatting the scaling (1D) tensors to be applied on the given argument tensors
+    # pads axis to (1D) tensors to then be broadcasted
+    size1 = [1] * len(weight1.size())
+    size1[output_axis] = weight1.size(output_axis)
+    size2 = [1] * len(weight2.size())
+    size2[input_axis] = weight2.size(input_axis)
+
+    scaling_factors = torch.reshape(scaling_factors, size2)
+    inverse_scaling_factors = torch.reshape(inverse_scaling_factors, size1)
+
+    weight1 = weight1 * inverse_scaling_factors
+    weight2 = weight2 * scaling_factors
+
+    module1.weight = torch.nn.Parameter(weight1)
+    module1.bias = torch.nn.Parameter(bias)
+    module2.weight = torch.nn.Parameter(weight2)
+
+def equalize(model, paired_modules_list, threshold=1e-4, inplace=True):
+    ''' Given a list of adjacent modules within a model, equalization will
+    be applied between each pair, this will repeated until convergence is achieved
+
+    Keeps a copy of the changing modules from the previous iteration, if the copies
+    are not that different than the current modules (determined by converged_test),
+    then the modules have converged enough that further equalizing is not necessary
+
+    '''
+    if not inplace:
+        model = copy.deepcopy(model)
+
+    name_to_module = {}
+    previous_name_to_module: Dict[str, torch.nn.modules.module] = {}
+    name_set = {name for pair in paired_modules_list for name in pair}
+
+    for name, module in model.named_modules():
+        if name in name_set:
+            name_to_module[name] = module
+            previous_name_to_module[name] = None
+    while not converged(name_to_module, previous_name_to_module, threshold):
+        for pair in paired_modules_list:
+            previous_name_to_module[pair[0]] = copy.deepcopy(name_to_module[pair[0]])
+            previous_name_to_module[pair[1]] = copy.deepcopy(name_to_module[pair[1]])
+
+            cross_layer_equalization(name_to_module[pair[0]], name_to_module[pair[1]])
+
+    if not inplace:
+        return model
+
+def converged(curr_modules, prev_modules, threshold=1e-4):
+    ''' Tests for the summed norm of the differences between each set of modules
+    being less than the given threshold
+
+    Takes two dictionaries mapping names to modules, the set of names for each dictionary
+    should be the same, looping over the set of names, for each name take the differnce
+    between the associated modules in each dictionary
+
+    '''
+    if len(curr_modules) != len(prev_modules):
+        raise TypeError("incompatiable modules in convergence condition")
```

Comment:
nit: Maybe say that the sizes are not compatible? In case the user is confused what exactly is not compatible.